README.md
# perl-alien-geos
A Perl Alien package to install [GEOS, the Geometry Engine, Open Source](http://geos.osgeo.org/)

[![Build Status](https://travis-ci.org/shawnlaffan/perl-alien-geos.svg?branch=master)](https://travis-ci.org/shawnlaffan/perl-alien-geos)

[![Build status](https://ci.appveyor.com/api/projects/status/s4c5nmnrk0m1kpkr?svg=true)](https://ci.appveyor.com/project/shawnlaffan/perl-alien-geos)

[![CPAN version](https://badge.fury.io/pl/Alien-geos-af.svg)](http://badge.fury.io/pl/Alien-geos-af)


Changes
Revision history for Alien::geos::af

1.012 2023-05-18
    - Share builds: patch one file to build using gcc-13 on Windows.

1.011 2022-11-26
    - Share builds: Use https for downloads

1.010 2022-03-26
    - Share builds:
      - Set rpath in compilation so geos.so finds geos_c.so
      - Set ccache base dir to improve cache hits
      - Use multiple probes to find geos.pc, geos-config
        or (windows only) GIS Internals files
    - General alienfile cleanups 

1.009 2022-02-03
    - share builds:
      - Modify geos-config to dynamically return paths relative to install directory.
      - Store alien properties from pkgconfig file instead of geos-config 
        (earlier geos versions did not have a geos.pc file).

1.008 2021-09-11
    - Switch to cmake.  GEOS 3.10 will remove autotools support.

1.007 2019-01-06
    - Require FFI::CheckLib 0.23 so dynamic_libs calls find geos-3-7-0.dll
    - Ensure we use the correct geos-config when setting alien props
      (some systems tried using the non-executable versions)

1.005 2018-11-15
    - More updates to the dynamic_libs method to find
      geos and geos_c libs for shared installs on windows

1.004 2018-11-15
    - Update the dynamic_libs method to find geos and geos_c
      libs for shared installs on windows

1.003 2018-11-15
    - Provide our own dynamic_libs method so system
      installs return the correct libs.

1.002 2018-11-13
    - Add File::Find::Rule to CONFIGURE_REQUIRES.
      This should avoid some CAN testers fails.

1.001 2018-11-12
    - First release to cpan


Makefile.PL
use strict;
use warnings;
use 5.010;
use ExtUtils::MakeMaker;

my %common_reqs = (
    "Alien::Build" => "2.48",
    "ExtUtils::MakeMaker" => "6.52",
    "Alien::Build::MM" => "1.41",
    "File::Find::Rule" => 0,
    "PkgConfig" => 0.19026,
    "Alien::Build::Plugin::Cleanse::BuildDir" => 0,
    'Sort::Versions' => 0,
    ($^O =~ /mswin/i) ? ("Alien::MSYS" => 0) : (),
    'Alien::cmake3' => 0.05,
);

my %WriteMakefileArgs = (
    "ABSTRACT" => "Alien package for GEOS - Geometry Engine, Open Source",
    "AUTHOR"   => 'Shawn Laffan <shawnlaffan@gmail.com>',
    "NAME"     => "Alien::geos::af",
    "VERSION_FROM" => "lib/Alien/geos/af.pm",
    "CONFIGURE_REQUIRES" => {
        %common_reqs,
        "PkgConfig"  => 0,
    },
    "BUILD_REQUIRES" => {
        %common_reqs,
        "File::Find::Rule" => 0,
        "Path::Tiny" => 0,
        "Capture::Tiny" => 0,
    },
    "TEST_REQUIRES" => {
        "Test::Alien" => 0,
        "FFI::CheckLib" => "0.23",
    },
    "PREREQ_PM" => {
        "Alien::Base"   => "1.41",
        "FFI::CheckLib" => "0.23",
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/shawnlaffan/perl-alien-geos',
            },
            bugtracker => {
                web => 'https://github.com/shawnlaffan/perl-alien-geos/issues/',
            },
        },
    },
    "DISTNAME" => "Alien-geos-af",
    "LICENSE"  => "perl",
);

use Alien::Build::MM;
my $abmm = Alien::Build::MM->new (
  clean_install => 1,
);
sub MY::install {
  $abmm->mm_install(@_);
}

%WriteMakefileArgs = $abmm->mm_args(%WriteMakefileArgs);

WriteMakefile(%WriteMakefileArgs);

sub MY::postamble {
  $abmm->mm_postamble;
}


lib/Alien/geos.pm
package Alien::geos;

use strict;
use warnings;
use parent qw( Alien::Base );

our $VERSION = '1.001';


1;

__END__

=head1 NAME

Alien::geos - Compile GEOS, the Geometry Engine, Open Source

=head1 BUILD STATUS
 
=begin HTML
 
<p>
    <img src="https://img.shields.io/badge/perl-5.10+-blue.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/shawnlaffan/perl-alien-geos"><img src="https://travis-ci.org/shawnlaffan/perl-alien-geos.svg?branch=master" /></a>
    <a href="https://ci.appveyor.com/project/shawnlaffan/perl-alien-geos"><img src="https://ci.appveyor.com/api/projects/status/1tqk5rd40cv2ve8q?svg=true" /></a>
</p>

=end HTML

=head1 SYNOPSIS

    use Alien::geos;

    use Env qw(@PATH);
    unshift @PATH, Alien::geos->bin_dir;

    print Alien::geos->dist_dir;

    
=head1 DESCRIPTION

GEOS is the Geometry Engine, Open Source.  See L<http://geos.osgeo.org/>.


=head1 REPORTING BUGS

Please send any bugs, suggestions, or feature requests to 
L<https://github.com/shawnlaffan/perl-alien-geos/issues>.

=head1 SEE ALSO

L<Geo::GDAL>

L<Geo::GDAL::FFI>

=head1 AUTHORS

Shawn Laffan, E<lt>shawnlaffan@gmail.comE<gt>


=head1 COPYRIGHT AND LICENSE


Copyright 2018 by Shawn Laffan


This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


lib/Alien/geos/af.pm
package Alien::geos::af;

use strict;
use warnings;
use parent qw( Alien::Base );

our $VERSION = '1.012';

#  make sure we find geos and geos_c
sub dynamic_libs {
    my ($class) = @_;

    use FFI::CheckLib 0.23 ();

    if ($class->install_type('system')) {
        my @libs;
        push @libs, FFI::CheckLib::find_lib(
            lib => 'geos',
        );
        push @libs, FFI::CheckLib::find_lib(
            lib => 'geos_c',
        );
        return wantarray ? @libs : $libs[0];
    }
    else {
        my $dir = $class->dist_dir;

        my $dynamic = Path::Tiny->new($class->dist_dir, 'dynamic');
        if (-d $dynamic) {
            $dir = $dynamic;
        }

        my @libs = $class->SUPER::dynamic_libs;

        push @libs, FFI::CheckLib::find_lib(
            lib        => ['geos', 'geos_c'],
            libpath    => $dir,
            systempath => [],
            recursive  => 1,
        );
        #warn "FOUND LIBS: " . join (':', @libs);
        return wantarray ? @libs : $libs[0];
    }
}

sub cflags {
    my ($self) = @_;
    $self->SUPER::cflags;
}

1;

__END__

=head1 NAME

Alien::geos::af - Compile GEOS, the Geometry Engine, Open Source

=head1 BUILD STATUS
 
=begin HTML
 
<p>
    <img src="https://img.shields.io/badge/perl-5.10+-blue.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/shawnlaffan/perl-alien-geos"><img src="https://travis-ci.org/shawnlaffan/perl-alien-geos.svg?branch=master" /></a>
    <a href="https://ci.appveyor.com/project/shawnlaffan/perl-alien-geos"><img src="https://ci.appveyor.com/api/projects/status/1tqk5rd40cv2ve8q?svg=true" /></a>
</p>

=end HTML

=head1 SYNOPSIS

    use Alien::geos::af;

    use Env qw(@PATH);
    unshift @PATH, Alien::geos::af->bin_dir;

    print Alien::geos::af->dist_dir;

    
=head1 DESCRIPTION

GEOS is the Geometry Engine, Open Source.  See L<http://geos.osgeo.org/>.

The name Aien::geos::AF is chosen to not clash with a pre-existing
Alien::GEOS distribution.  This package differs in that it uses the
alienfile approach, hence the ::af suffix in the name.


=head1 REPORTING BUGS

Please send any bugs, suggestions, or feature requests to 
L<https://github.com/shawnlaffan/perl-alien-geos/issues>.

=head1 SEE ALSO

L<Geo::GDAL::FFI>

L<Alien::gdal>

L<Alien::proj>

L<Geo::Geos>

=head1 AUTHORS

Shawn Laffan, E<lt>shawnlaffan@gmail.comE<gt>


=head1 COPYRIGHT AND LICENSE


Copyright 2018- by Shawn Laffan


This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut
